File: java/src/main/java/sn/prime/HMACService.java
package sn.prime;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.Map;

/**
 * HMAC Verification Service for validating webhook requests using SHA-512.
 */
public class HMACService {
    private final String hmacKey;
    private final String apiKey;

    /**
     * Creates a new HMAC Service instance with provided HMAC_KEY and API_KEY.
     *
     * @param hmacKey The secret key used for HMAC signature generation
     * @param apiKey  The API key for request authentication
     * @throws IllegalArgumentException if required configuration is missing
     */
    public HMACService(String hmacKey, String apiKey) {
        if (hmacKey == null || hmacKey.isEmpty() || apiKey == null || apiKey.isEmpty()) {
            throw new IllegalArgumentException("HMAC key and API key are required");
        }
        this.hmacKey = hmacKey;
        this.apiKey = apiKey;
    }

    /**
     * Verifies request headers and payload signature.
     *
     * @param headers Request headers containing X-Api-Key and X-Hmac-Signature
     * @param payload Request payload as a Payload object
     * @return VerificationResult containing validation status and error message if any
     */
    public VerificationResult verify(Map<String, String> headers, Payload payload) {
        try {
            // Check API key
            String providedApiKey = headers.get("x-api-key");
            if (providedApiKey == null || !providedApiKey.equals(this.apiKey)) {
                return new VerificationResult(false, "Invalid API key");
            }

            // Check HMAC signature
            String providedSignature = headers.get("x-hmac-signature");
            if (providedSignature == null) {
                return new VerificationResult(false, "Missing HMAC signature");
            }

            String calculatedSignature = calculateSignature(payload);
            if (!providedSignature.equals(calculatedSignature)) {
                return new VerificationResult(false, "Invalid signature");
            }

            return new VerificationResult(true, null);
        } catch (Exception e) {
            return new VerificationResult(false, "Verification failed: " + e.getMessage());
        }
    }

    /**
     * Calculates HMAC signature for the given payload using SHA-512.
     *
     * @param payload Request payload as a Payload object
     * @return Base64 encoded HMAC signature
     * @throws RuntimeException if signature calculation fails
     */
    public String calculateSignature(Payload payload) {
        try {
            // Validate required fields
            validatePayload(payload);

            // Build data string
            String dataToSign = payload.statusPayment() +
                    payload.primeClientPhone() +
                    payload.amount();

            // Create HMAC SHA-512 instance
            Mac sha512Hmac = Mac.getInstance("HmacSHA512");
            SecretKeySpec secretKey = new SecretKeySpec(
                    hmacKey.getBytes(StandardCharsets.UTF_8),
                    "HmacSHA512"
            );
            sha512Hmac.init(secretKey);

            // Calculate HMAC
            byte[] hmacBytes = sha512Hmac.doFinal(dataToSign.getBytes(StandardCharsets.UTF_8));

            // Encode to Base64
            return Base64.getEncoder().encodeToString(hmacBytes);
        } catch (Exception e) {
            throw new RuntimeException("Failed to calculate signature: " + e.getMessage());
        }
    }

    /**
     * Validates the payload for required fields.
     *
     * @param payload Request payload as a Payload object
     * @throws IllegalArgumentException if any required field is missing or invalid
     */
    private void validatePayload(Payload payload) {
        if (payload == null) {
            throw new IllegalArgumentException("Payload cannot be null");
        }

        if (payload.statusPayment() == null || payload.statusPayment().isEmpty()) {
            throw new IllegalArgumentException("Missing required field: statusPayment");
        }

        if (payload.primeClientPhone() == null || payload.primeClientPhone().isEmpty()) {
            throw new IllegalArgumentException("Missing required field: primeClientPhone");
        }

        if (payload.amount() == null || payload.amount().isEmpty()) {
            throw new IllegalArgumentException("Missing required field: amount");
        }
    }

    /**
     * Result class for verification operations.
     */
    public record VerificationResult(boolean isValid, String error) {
        @Override
        public String toString() {
            return "VerificationResult{" +
                    "isValid=" + isValid +
                    (error != null ? ", error='" + error + "'" : "") +
                    '}';
        }
    }
}

File: javascript/src/HMACService.js
/**
 * @fileoverview HMAC Verification SDK
 * This SDK provides HMAC-based request verification functionality for secure API communications.
 * It implements HMAC-SHA512 signature generation and verification along with API key validation.
 * @version 1.0.1
 */
import crypto from "crypto";

/**
 * Configuration options for the HMACService
 * @typedef {Object} HMACConfig
 * @property {string} hmacKey - The secret key used for HMAC signature generation
 * @property {string} apiKey - The API key for request authentication
 */

/**
 * Response structure for verification results
 * @typedef {Object} VerificationResult
 * @property {boolean} isValid - Indicates if the verification was successful
 * @property {string} [error] - Error message if verification failed
 */

/**
 * Payload structure expected for signature calculation
 * @typedef {Object} Payload
 * @property {string} statusPayment - Payment status
 * @property {string} primeClientPhone - Client phone number
 * @property {number} amount - Transaction amount
 */

export class HMACService {
  #hmacKey;
  #apiKey;

  /**
   * Creates an instance of HMACService
   * @param {HMACConfig} config - Configuration object
   * @throws {Error} If required configuration parameters are missing
   */
  constructor(config) {
    this.#validateConfig(config);
    this.#hmacKey = config.hmacKey;
    this.#apiKey = config.apiKey;
  }

  /**
   * Validates the configuration object
   * @private
   * @param {HMACConfig} config - Configuration object to validate
   * @throws {Error} If validation fails
   */
  #validateConfig(config) {
    if (!config?.hmacKey || typeof config.hmacKey !== "string") {
      throw new Error("HMAC key is required and must be a string");
    }
    if (!config?.apiKey || typeof config.apiKey !== "string") {
      throw new Error("API key is required and must be a string");
    }
  }

  /**
   * Calculates HMAC signature for the given payload using SHA512
   * @param {Payload} payload - The payload to sign
   * @returns {string} Base64 encoded HMAC signature
   * @throws {Error} If signature calculation fails
   */
  calculateSignature(payload) {
    try {
      this.#validatePayload(payload);

      // Create HMAC instance with SHA512
      const hmac = crypto.createHmac(
        "sha512",
        Buffer.from(this.#hmacKey, "utf8")
      );

      // Build and sign data string
      const dataToSign = this.#buildDataString(payload);
      hmac.update(Buffer.from(dataToSign, "utf8"));

      return hmac.digest("base64");
    } catch (error) {
      throw new Error(`HMAC calculation failed: ${error.message}`);
    }
  }

  /**
   * Validates the payload structure
   * @private
   * @param {Payload} payload - Payload to validate
   * @throws {Error} If payload validation fails
   */
  #validatePayload(payload) {
    const requiredFields = ["statusPayment", "primeClientPhone", "amount"];
    const missingFields = requiredFields.filter((field) => !payload?.[field]);
    if (missingFields.length > 0) {
      throw new Error(`Missing required fields: ${missingFields.join(", ")}`);
    }
  }

  /**
   * Builds the data string for signature calculation
   * @private
   * @param {Payload} payload - Payload to process
   * @returns {string} Concatenated data string
   */
  #buildDataString(payload) {
    return `${payload.statusPayment}${payload.primeClientPhone}${payload.amount}`;
  }

  /**
   * Verifies request headers and payload signature
   * @param {Object.<string, string>} headers - Request headers
   * @param {Payload} payload - Request payload
   * @returns {VerificationResult} Verification result
   */
  verify(headers, payload) {
    try {
      // Normalize and validate headers
      const normalizedHeaders = this.#normalizeHeaders(headers);

      // API key verification
      if (normalizedHeaders["x-api-key"] !== this.#apiKey) {
        return {
          isValid: false,
          error: "Invalid API key",
        };
      }

      // Signature verification
      const providedSignature = normalizedHeaders["x-hmac-signature"];
      const calculatedSignature = this.calculateSignature(payload);

      if (providedSignature !== calculatedSignature) {
        return {
          isValid: false,
          error: "Invalid signature",
        };
      }

      return { isValid: true };
    } catch (error) {
      return {
        isValid: false,
        error: error.message,
      };
    }
  }

  /**
   * Normalizes header keys to lowercase
   * @private
   * @param {Object.<string, string>} headers - Headers to normalize
   * @returns {Object.<string, string>} Normalized headers
   */
  #normalizeHeaders(headers) {
    return Object.entries(headers).reduce((acc, [key, value]) => {
      acc[key.toLowerCase()] = value;
      return acc;
    }, {});
  }
}
